toolchain/parser/parser_context.h

```diff
@@ -379,6 +296,29 @@ class Parser {
   llvm::SmallVector<StateStackEntry> state_stack_;
 };
 
+// `clang-format` has a bug with spacing around `->` returns in macros. See
+// https://bugs.llvm.org/show_bug.cgi?id=48320 for details.
+#define CARBON_PARSER_STATE(Name) \
+  auto ParserHandle##Name(ParserContext& context)->void;
+#include "toolchain/parser/parser_state.def"
+
+// The diagnostics below may be emitted a couple different ways as part of
+// operator parsing.
+
+CARBON_DIAGNOSTIC(
+    OperatorRequiresParentheses, Error,
+    "Parentheses are required to disambiguate operator precedence.");
+
+CARBON_DIAGNOSTIC(ExpectedSemiAfterExpression, Error,
+                  "Expected `;` after expression.");
+
+CARBON_DIAGNOSTIC(ExpectedDeclarationName, Error,
+                  "`{0}` introducer should be followed by a name.", TokenKind);
+CARBON_DIAGNOSTIC(ExpectedDeclarationSemiOrDefinition, Error,
+                  "`{0}` should either end with a `;` for a declaration or "
+                  "have a `{{ ... }` block for a definition.",
+                  TokenKind);
+
```

Comment:
This section feels a bit awkward in `parse_context.h`. What do you think about pulling these into a `parser_handle.h` or similar so this can be a bit more focused on just the shared *context*, but not the handlers and dispatching?

Reply:
I'd started with that, but:

1) I know I need to clean up the diagnostics, but a helper function might be a better approach.
2) If I address the first, then we end up with a header file that's 3 lines of code (the macro). Everything that includes that file must also include this file.

It felt weird adding headers with 3 LOC.

Reply:
I mean, 3 LOC before the xmacro header, then it'll be a *huge* header...

But I see the point that there isn't much saved by having it split up at this point, so yeah, leaving all of this in one file for now makes lots of sense.

What do you think about going back to `parser.h` as a file name, especially given the above comment on it really handling *anything* that shouldn't go into more split-up header files? I'm happy having the implementation split into parser_context.cpp (or not). I'd also be happy w/ the tiny bit of code for implementing the Parser function staying in parse_tree.cpp or moving to a parser.cpp. I'd probably have both parser_context.cpp and parser.cpp left to my own devices, but I don't feel strongly at all.

Reply:
SGTM.

Regarding parser.h, my preference is to name files after the main contained object when there is one, which to me is the ParserContext. While there are all the handlers, they are handlers for the context, so I don't view them as in conflict. For me, this is the strongest reason for a parser_handler.h -- to mirror the split cpp files. Regardless, the declaration code is minimal and I don't think it helps readers. To the point of the actual header size, maybe it would help compilation, but there may be better fixes for compilation, depending on what we see.

But I see a more significant issue with parser.h: it'd be consistent to also rename semantics_context -> semantics, lowering_context -> lowering in #2819 and #2820. However, while all of these would be eponymous with the directory, none of them would be the entry point:

- ParseTree::Parse in parse_tree.h
- SemanticsIR::MakeFromParseTree in semantics_ir.h
- LowerToLLVM in lower_to_llvm.h

I think naming targets eponymously with directories is best done with external APIs rather than internal libraries; that's where I would always look first. It could actually be confusing if we have all these eponymously named headers with partial functionality, but the actual entry point is in some arbitrary header.

Reply:
(definitely not blocking, just continuing discussion here)

> Regarding parser.h, my preference is to name files after the main contained object when there is one, which to me is the ParserContext. While there are all the handlers, they are handlers for the context, so I don't view them as in conflict. For me, this is the strongest reason for a parser_handler.h -- to mirror the split cpp files. Regardless, the declaration code is minimal and I don't think it helps readers. To the point of the actual header size, maybe it would help compilation, but there may be better fixes for compilation, depending on what we see.
> 
> But I see a more significant issue with parser.h: it'd be consistent to also rename semantics_context -> semantics, lowering_context -> lowering in #2819 and #2820. However, while all of these would be eponymous with the directory, none of them would be the entry point:
> 
> * ParseTree::Parse in parse_tree.h
> * SemanticsIR::MakeFromParseTree in semantics_ir.h
> * LowerToLLVM in lower_to_llvm.h
> 
> I think naming targets eponymously with directories is best done with external APIs rather than internal libraries; that's where I would always look first. It could actually be confusing if we have all these eponymously named headers with partial functionality, but the actual entry point is in some arbitrary header.

Yeah, this is a good point about matching the directory name.

FWIW, I somewhat can see both semantics and lowering wanting their own nomenclature anyways here, to reflect some important differences. (Similar to how their directory names don't exactly match, and for similar reasons.)

Anyways, if there are good naming conventions here, seems better to search for them in follow-up PRs and not stress here.